src/apify/scrapy/middlewares/apify_proxy.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import annotations
+
+from base64 import b64encode
+from typing import TYPE_CHECKING
+from urllib.parse import ParseResult, unquote, urlparse
+
+from scrapy.core.downloader.handlers.http11 import TunnelError
+from scrapy.exceptions import NotConfigured
+from scrapy.utils.python import to_bytes
+
+from apify import Actor, ProxyConfiguration
+
+if TYPE_CHECKING:
+    from scrapy import Request, Spider
+    from scrapy.crawler import Crawler
+
+
+class ApifyHttpProxyMiddleware:
+    """Apify HTTP proxy middleware for Scrapy.
+
+    This middleware enhances request processing by adding a 'proxy' field to the request's meta and an
+    authentication header. It draws inspiration from the Scrapy default `HttpProxyMiddleware`. The proxy URL is
+    sourced from the settings under the `APIFY_PROXY_SETTINGS` key. The value of this key, a dictionary, should be
+    provided by the Actor input. An example of the proxy settings:
+
+    proxy_settings = {'useApifyProxy': true, 'apifyProxyGroups': []}
+    """
+
+    def __init__(
+        self: ApifyHttpProxyMiddleware,
+        proxy_settings: dict,
+        auth_encoding: str = 'latin-1',
```

Comment:
Is there a realistic usecase for overriding this?

Reply:
Scrapy [HttpProxyMiddleware](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/httpproxy.py#L25C66-L25C66) handles the `auth_encoding` in the same way, so I stick with that.

Would you rather have it as a module/class level constant?

Reply:
I moved the `get_basic_auth_header` to a separate module, so the optional argument is moved just to that function.

Reply:
And if you decide to use that argument, can you comfortably set the header value in the ApifyHttpProxyMiddleware class?

By the way, `httpx` for example handles this by accepting both `str` and `bytes` for `username` and `password`, and if you want to control the encoding, you just pass pre-encoded `bytes`

Reply:
Discussed on Slack... We're gonna leave it as it is for now, if someone in the future needs to set up the encoding, we can provide it by extracting optional value from the `crawler.settings` in the `from_crawler` method and pass it to the default constructor.